constantine/commitments/ipa/multiproof.nim

```diff
@@ -0,0 +1,337 @@
+            var scaledEvals {.noInit.}: EC_P_Fr
+            scaledEvals.prod(r, Fs[i][j])
+            groupedFs[z][j].sum(groupedFs[z][j], scaledEvals)
+        
+    
+    var gx : array[DOMAIN, EC_P_Fr]
+
+    for idx in 0..<DOMAIN:
+        if groupedFs[idx].len == 0:
+            continue
+
+        var quotient: array[DOMAIN,EC_P_Fr]
+        var passer : int
+        passer = idx
+        quotient.divisionOnDomain(precomp, passer, groupedFs[idx])
+
+        for j in  0..<DOMAIN:
+            gx[j] += quotient[j]
+        
+    var D: EC_P
+    D.pedersen_commit_varbasis(basis,basis.len, gx, gx.len)
+
+    transcript.pointAppend(asBytes"D", D)
+
+    var t {.noInit.}: matchingOrderBigInt(Banderwagon)
+    t.generateChallengeScalar(transcript,asBytes"t")
+
+    var t_fr {.noInit.}: EC_P_Fr
+    t_fr.fromBig(t)
+
+    # Computing the denominator inverses only for referenced evaluation points.
+    var denInv {.noInit.}: array[DOMAIN, EC_P_Fr]
+    for i in 0..<DOMAIN:
+        denInv[i].setZero()
+
+    for z in 0..<DOMAIN:
+        if groupedFs[z].len == 0:
+            continue
+
+        var z_fr {.noInit.} : EC_P_Fr
+        z_fr.domainToFrElem(uint8(z))
+        var deno {.noInit.}: EC_P_Fr
+
+        deno.diff(t_fr,z_fr)
+        var idxx = 0
+        denInv[idxx] = deno
+        idxx = idxx + 1
+
+
+    var denInv_prime {.noInit.} : array[DOMAIN, EC_P_Fr]
+    denInv_prime.batchInvert(denInv)
+
+    #Compute h(X) = g1(X)
+    var hx {.noInit.}: array[DOMAIN, EC_P_Fr]
+    var denInvIdx = 0
+
+    for i in 0..<DOMAIN:
+        if groupedFs[i].len == 0:
+            continue
+
+        for k in 0..<DOMAIN:
+            var tmp {.noInit.}: EC_P_Fr
+            tmp.prod(groupedFs[i][k], denInv[denInvIdx])
+            hx[k].sum(hx[k], tmp)
+
+        denInvIdx = denInvIdx + 1
+
+    var hMinusg {.noInit.}: array[DOMAIN, EC_P_Fr]
+
+    for i in 0..<DOMAIN:
+        hMinusg[i].diff(hx[i],gx[i])
+
+    var E: EC_P
+
+    E.pedersen_commit_varbasis(basis,basis.len, hx, hx.len)
+    transcript.pointAppend(asBytes"E",E)
+
+    var EMinusD: EC_P
+
+    EMinusD.diff(E,D)
+
+    var ipaProof: IPAProof
+
+    var checks: bool
+    checks = ipaProof.createIPAProof(transcript, ipaSetting, EMinusD, hMinusg, t_fr)
+
+    doAssert checks == true, "Could not compute IPA Proof!"
+
+    res.IPAprv = ipaProof
+    res.D = D
+    success = true
+
+# ############################################################
+#
+#                 Multiproof Verification
+#
+# ############################################################
+    
+# Mutliproof verifier verifies the multiproof for several polynomials in the evaluation form
+# The list of triplets (C,Y, Z) represents each polynomial commitment, evaluation
+# result, and evaluation point in the domain 
+func verifyMultiproof* [bool] (res: var bool, transcript : var sha256, ipaSettings: IPASettings, multiProof: var MultiProof, Cs: openArray[EC_P], Ys: openArray[EC_P_Fr], Zs: openArray[uint8])=
```

Comment:
```suggestion
func verifyMultiproof*(transcript : var sha256, ipaSettings: IPASettings, multiProof: var MultiProof, Cs: openArray[EC_P], Ys: openArray[EC_P_Fr], Zs: openArray[uint8]): bool =
```

If it's a bool, it should be returned as a bool

Reply:
Done